### PR TITLE
Remove a back-compat hack on lazy TAIT

### PR DESCRIPTION
**Pull request by JohnTitor**
This PR's motivation is here: https://github.com/rust-lang/rust/issues/72614#issuecomment-1134595446
~~But removing a hack doesn't seem to reject the code on the issue, there're some more hacks?~~
r? @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
> But removing a hack doesn't seem to reject the code on the issue, there're some more hacks?

Oh... Turns out I didn't read your example test very well. Yea, that's totally fine to pass. There were other roundabout ways to make the same thing pass before.

So independently of that issue, we can still remove the back compat hack and see how well/bad it goes

**Comment by JohnTitor** [comment]
Alright, dropped the test as it's unrelated to the hack itself.

> Oh... Turns out I didn't read your example test very well. Yea, that's totally fine to pass. There were other roundabout ways to make the same thing pass before.

Is there any reference/discussion etc. to that? To close #72614, I'd like to know how these impl traits work!

> So independently of that issue, we can still remove the back compat hack and see how well/bad it goes

Feel free to kick off crater once you're ready, I'm not familiar with it.

File: compiler/rustc_infer/src/infer/opaque_types.rs [hunk 1]
```diff
@@ -39,20 +39,15 @@ pub struct OpaqueTypeDecl<'tcx> {
...
         &self,
-        value: T,
+        ty: Ty<'tcx>,
         body_id: HirId,
         span: Span,
         param_env: ty::ParamEnv<'tcx>,
-    ) -> InferOk<'tcx, T> {
-        if !value.has_opaque_types() {
-            return InferOk { value, obligations: vec![] };
-        }
```
Comment (oli-obk) [inline compiler/rustc_infer/src/infer/opaque_types.rs:53]: This may still be necessary as a performance optimization

**Comment by oli-obk** [comment]
> Is there any reference/discussion etc. to that? To close #72614, I'd like to know how these impl traits work!

Not really, but I'll leave a comment on that issue explaining what the heck is going on

File: src/test/ui/impl-trait/nested-return-type2.rs [hunk 2]
```diff
@@ -23,6 +21,7 @@ impl<R: Duh, F: FnMut() -> R> Trait for F {
 // Lazy TAIT would error out, but we inserted a hack to make it work again,
```
Comment (mati865) [inline src/test/ui/impl-trait/nested-return-type2.rs:21]: Rember to remove this comment if crater result is acceptable.

**Comment by JohnTitor** [comment]
@oli-obk We have a ton of spurious-regressed, is it okay?

Some analysis of "root results" failures:
- stuartZhang.scaffold-wizard.ae8617257ead4fb6fc1e529f192fc4ccbe3806c6: Seems unrelated
- Ayrx.binja-rs-hello-world.a0f5fb90a3544b233856cddbdac059da21ce27ee: Seems unrelated
- ledger-transport-zemu-0.9.0: Seems unrelated
- bitrust-0.1.0: Seems unrelated
- andrewhickman.proxy-server.1f02e7d8fafb8f43059ae0313f5dc3ca64cee637: Seems **related** but the crate has not been updated since 2020 Sep. 8th
- battjt.cb-test.b23d7048a84aa9f1e407edb24fddc001794e31d6: Seems unrelated, the issue has been reported as #88643 and it's fixed on master

I think all of the "regressed: dependencies" failures are due to a lack of mem.
So I'd like to say that the impact of this change would be pretty small.
Thoughts?

**Comment by oli-obk** [comment]
Yea, except for proxy-server, all of these are unrelated. I guess we can break them as long as we open an issue or fix PR on their repo

**Comment by JohnTitor** [comment]
Submitted https://github.com/andrewhickman/mock-server/pull/1 and addressed review comments.

**Comment by jackh726** [comment]
@rfcbot reviewed

**Comment by oli-obk** [comment]
r=me after a rebase

**Pull request merged**